drivers/gpio/gpio_eos_s3.c

```diff
@@ -0,0 +1,432 @@
+		case GPIO_0:
+			cfg->xPadConf->ucPin = PAD_24;
+			cfg->xPadConf->ucFunc = PAD24_FUNC_SEL_GPIO_0;
+			break;
+		case GPIO_1:
+			cfg->xPadConf->ucPin = PAD_26;
+			cfg->xPadConf->ucFunc = PAD26_FUNC_SEL_GPIO_1;
+			break;
+		case GPIO_2:
+			cfg->xPadConf->ucPin = PAD_28;
+			cfg->xPadConf->ucFunc = PAD28_FUNC_SEL_GPIO_2;
+			break;
+		case GPIO_3:
+			cfg->xPadConf->ucPin = PAD_30;
+			cfg->xPadConf->ucFunc = PAD30_FUNC_SEL_GPIO_3;
+			break;
+		case GPIO_4:
+			cfg->xPadConf->ucPin = PAD_31;
+			cfg->xPadConf->ucFunc = PAD31_FUNC_SEL_GPIO_4;
+			break;
+		case GPIO_5:
+			cfg->xPadConf->ucPin = PAD_36;
+			cfg->xPadConf->ucFunc = PAD36_FUNC_SEL_GPIO_5;
+			break;
+		case GPIO_6:
+			cfg->xPadConf->ucPin = PAD_38;
+			cfg->xPadConf->ucFunc = PAD38_FUNC_SEL_GPIO_6;
+			break;
+		case GPIO_7:
+			cfg->xPadConf->ucPin = PAD_45;
+			cfg->xPadConf->ucFunc = PAD45_FUNC_SEL_GPIO_7;
+			break;
+		default:
+			return -EINVAL;
+		}
+	} else {
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+/*
+ * NOTE:
+ * Pin number defined in the device tree should be actually a GPIO number
+ * to avoid overflow in high level GPIO API.
+ * `gpio_num_extended` in range of 0 to 7 matches to `port_num == 0` PADs from
+ * `find_pad` function and `gpio_num_extended` in range of 8 to 15
+ * are associated with `port_num == 1` consequently.
+ */
+static int gpio_eos_s3_configure(const struct device *dev,
+				 gpio_pin_t gpio_num_extended,
+				 gpio_flags_t flags)
+{
+	ARG_UNUSED(dev);
+	int ret;
+	GPIOCfgTypeDef gpio_cfg;
+	PadConfig pad_conf;
+	int port_num = 0;
+
+	gpio_cfg.xPadConf = &pad_conf;
+
+	/* Check for an invalid gpio number */
+	if (gpio_num_extended >= MAX_GPIOS) {
+		return -EINVAL;
+	}
+
+	/* Determine "port" number */
+	if ((int)(gpio_num_extended / MAX_GPIOS_PER_PORT) == 1) {
+		port_num = 1;
+	}
+
+	/* Get actual GPIO number */
+	uint8_t gpio_num = gpio_num_extended % 8;
+
+	/* Check for an invalid pin configuration */
+	if (((flags & GPIO_INPUT) != 0) && ((flags & GPIO_OUTPUT) != 0)) {
+		return -ENOTSUP;
+	}
+
+	gpio_cfg.ucGpioNum = gpio_num;
+	ret = find_pad(&gpio_cfg, port_num);
+
+	if (ret) {
+		return ret;
+	}
+
+	/* Configure PAD */
+	if (flags & GPIO_PULL_UP) {
+		gpio_cfg.xPadConf->ucPull = PAD_PULLUP;
+	} else if (flags & GPIO_PULL_DOWN) {
+		gpio_cfg.xPadConf->ucPull = PAD_PULLDOWN;
+	} else {
+		gpio_cfg.xPadConf->ucPull = PAD_NOPULL;
+	}
+
+	if ((flags & GPIO_DIR_MASK) == GPIO_INPUT) {
+		gpio_cfg.xPadConf->ucMode = PAD_MODE_INPUT_EN;
+		gpio_cfg.xPadConf->ucSmtTrg = PAD_SMT_TRIG_EN;
+		HAL_PAD_Config(gpio_cfg.xPadConf);
+	} else {
+		gpio_cfg.xPadConf->ucMode = PAD_MODE_OUTPUT_EN;
+		HAL_PAD_Config(gpio_cfg.xPadConf);
+	}
+
+	return 0;
+}
+
+static int gpio_eos_s3_port_get_raw(const struct device *dev,
+				    uint32_t *value)
+{
+	ARG_UNUSED(dev);
+	uint8_t pin_value = 0;
+	*value = 0;
+
+	for (uint8_t gpion = 0; gpion < MAX_GPIOS_PER_PORT; gpion++) {
+		HAL_GPIO_Read(gpion, &pin_value);
+		*value |= (pin_value == 1) ? 1 << gpion : 0;
```

Comment:
Might be better to have a `gpio_port_value_t mask = 1U` that gets shifted up in the loop body, but if you want to do `1 << x` please use `BIT(x)` from `<sys/utils.h>`.

Reply:
thanks, I switched it to `BIT(x)`